### ISSUE TITLE
kd_loss implementation issue

### ISSUE THREAD

**Issue by ZaberKo** [HikariTJU/LD#74]
Hello, I found that the `knowledge_distillation_kl_div_loss()` in `mmdet/models/losses/kd_loss.py` uses a different implementation compared to the normal KL Div definition, which is equivalent to `F.kl_div(reduction='mean')` instead of `F.kl_div(reduction='batchmean')` as mentioned in [F.kl_div](https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html#torch.nn.KLDivLoss).
```
kd_loss = F.kl_div(
    F.log_softmax(pred / T, dim=1), target, reduction='none').mean(1) * (
        T * T)
```

The correct KL Div should be like
```
kd_loss = F.kl_div(
    F.log_softmax(pred / T, dim=1), target, reduction='none').sum(1) * (
        T * T)
```

Is there any reason to use the above implementation? Current kl_div is 1/17 smaller than the real kl_div, when gfl reg_max=16.

**Comment by HikariTJU** [comment]
I remeber that .mean(1) is equal to reduction='batch_mean‘ ？

**Comment by ZaberKo** [comment]
> I remeber that .mean(1) is equal to reduction='batch_mean‘ ？

Here is the source code of `F.kl_div`:
https://github.com/pytorch/pytorch/blob/defa0d3a2d230e5d731d5c443c1b9beda2e7fd93/torch/nn/functional.py#L2949-L2958

And the problem here is that the `kd_loss` is subsequently averaged by `@weighted_loss` wrapper.

**Comment by HikariTJU** [comment]
So batch_mean equals .mean(0)?

**Comment by ZaberKo** [comment]
> So batch_mean equals .mean(0)?

"batchmean" means .sum()/batch_size

**Comment by HikariTJU** [comment]
OK, I get your point, you mean mathmatically .sum(1) is the correct implementation and .mean(1)=.sum(1)/16
That's true.

**Comment by ZaberKo** [comment]
> OK, I get your point, you mean mathmatically .sum(1) is the correct implementation and .mean(1)=.sum(1)/16 That's true, but how is it related to batchmean?

BTW, I also found that `loss_ld` used weighted sum and was not divided by `avg_factor` (i.e. sum of weights)? Is this a typo or intended behavior for not using normalization?

**Comment by ZaberKo** [comment]
FYI: I record the factor ratio `avg_factor/(self.reg_max+1)` during the training. Maybe it will help this discussion.

<img width="741" alt="image" src="https://github.com/HikariTJU/LD/assets/20830726/ec7f45ed-d476-49bb-a055-c9a57cca224d">

**Comment by HikariTJU** [comment]
It's a intended behavior because experiment shows not dividing is better. Don't know the theory behind this though

**Comment by ZaberKo** [comment]
> It's a intended behavior because experiment shows not dividing is better. Don't know the theory behind this though

I see, thanks for the reply.
